python/ray/serve/tests/test_fastapi.py

```diff
@@ -490,6 +490,35 @@ class TestDeployment:
     assert resp.json() == {"a": "a", "b": ["b"]}
 
 
+def test_fastapiwrapper_constructor_before_startup_hooks(serve_instance):
+    """
+    Tests that the class constructor is called before the startup hooks
+    are run in `FastAPIWrapper`.
+    """
+    app = FastAPI()
+
+    @serve.deployment(route_prefix="/")
+    @serve.ingress(app)
+    class TestDeployment:
+        def __init__(self):
+            try:
+                # `self.lifespan` is defined in `FastAPIWrapper` before the
+                # startup hooks are called. If it is already initialized, the
+                # class constructor is not called before the startup hooks
+                self.lifespan
+                self.test_passed = False
```

Comment:
can we test the actual behavior instead of testing AttributeError? something like define an init method that will set event in the signal actor, and have the startup hook await for event in the signal actor. the process will only work with this kind of ordering.

```
from ray.test_utils import SignalActor
```

<br class="Apple-interchange-newline">

Reply:
Yeah, that's a good idea. I set the event in the startup hook and awaited in the init method because the startup hook cannot take in `self` and I wasn't able to figure out another way to return whether the timeout occurred from the startup hook to the main program. If an error occurs in the startup hook the app still finishes initializing.